### ISSUE TITLE
Script inline python requirement not respected when .python-version file is in project

### ISSUE THREAD

**Issue by Magglish** [astral-sh/uv#9393]
Hello,

I'm really exicted that we can specify python version and libraries dependency for scripts in `uv` and this thing can convinced me to move from `poetry` to `uv`.

**Description**

Unfortunately, I've encounter an issue which I don't if it's a bug or expected behaviour.
The problem is that script inline python requirement is not respected by `uv` when `.python-version` file is specified in the same directory.

**Steps to reproduce:**

1. OS Version: Ubuntu 22.04 LTS
2. `uv` version: **0.5.4** (installed through official installation script as mentioned in documentation)

3. `.python-version` contains:

```
3.12.7
```

4. `example.py` script with dependencies:

```python
# /// script
# requires-python = "==3.11.5"
# dependencies = [
#   "rich==13.9.4",
# ]
# ///

import sys
import time
from rich.progress import track

for i in track(range(20), description="For example:"):
    time.sleep(0.05)
print(sys.version)
```

5. Create venv with `uv`

```bash
uv venv
```

6. Run `example.py`

```bash
uv run example.py
```

What we see:

1. warning: The Python request from `.python-version` resolved to Python 3.12.7, which is incompatible with the script's Python requirement: `==3.11.5`
2. `print(sys.version)` returns `3.12.7` but not `3.11.5`

What should we see:

The script should be run with python 3.11.5, but it's not.

**Alternative**

We can run `example.py` with

```bash
uv run --python 3.11.5 example.py
```

But this means that the inline metadata requirements are not being used at all.

Is this a bug or a feature?

**Comment by FishAlchemist** [comment]
It has a warning, so I personally think it's an expected behavior.
```
warning: The Python request from `.python-version` resolved to Python 3.12.7, which is incompatible with the script's Python requirement: `==3.11.5`
```
Personally, I believe that if the script has a specified version, we ought to adhere to it unless a different version is explicitly defined via the CLI.

**Comment by Magglish** [comment]
In my opinion, if the `uv` wants to respect the inline metadata requirements presented in the scripts then all requirements should be respected.

**Comment by charliermarsh** [comment]
Yeah I believe this is intentional. Another option would be to throw a hard error because you have conflicting requirements.

**Comment by zanieb** [comment]
I think if the `.python-version` conflicts with a script's Python requirement, we should prefer the script's definition. This is consistent with projects which won't use a `.python-version` file from a parent directory if it conflicts with the project's requirement.

**Comment by charliermarsh** [comment]
So in that case, we would respect `.python-version` if it's consistent, and ignore it otherwise? (Not disagreeing, just confirming.)

**Comment by Magglish** [comment]
@zanieb did your team had a chance to talk about this issue and decide? Is there a chance that you will revert the implementation and the script inline requirements for python version will be respected? :)

**Comment by Magglish** [comment]
Bump

**Comment by ckutlu** [comment]
Bump.

I ran into this while writing a CLI helper for a library I'm working on. The library has to support python 3.9 while the CLI helper, being completely decoupled from the project apart from being hosted in the same repo, needs to be >3.9 because of a bug in one of its dependencies for older python versions.

**Comment by coezbek** [comment]
The documentation on this is currently misleading (see https://docs.astral.sh/uv/guides/scripts/#declaring-script-dependencies). It says:

```
uv also respects Python version requirements:

example.py

# /// script
# requires-python = ">=3.12"
# dependencies = []
# ///

# Use some syntax added in Python 3.12
type Point = tuple[float, float]
print(Point)

Note: The dependencies field must be provided even if empty.

uv run will search for and use the required Python version. The Python version will download if it is not 
installed — see the documentation on [Python versions](https://docs.astral.sh/uv/concepts/python-versions/) 
for more details.
```

As far as I can see, `uv` doesn't search and use the required Python version, but rather it uses the version specified in `.python-version` in the current or all parent folders and just throws a warning if there is a mismatch.

If `.python-version` is missing then the resolution works as expected and documented.

Maybe to summarize what I think should be the requirements of python version resolution when running scripts:

- If a script contains a version requirement, then this requirement is fulfilled or the run is aborted.
- If a `.python-version` is available and it matches the `requires-python`, then this version of python is used to run the script.
- If no `.python-version` is available, the lowest (?) version fulfilling the requirement which is already installed is used.
- If no python is installed fulfilling the requirement, `uv` will download the lowest (?) version fulfilling the `requires-python` version.
